### DISCUSSION TITLE
CAST3M Implementation - EKMC model

### DISCUSSION THREAD

**Discussion by RonanRiverie** [thelfer/tfel#576] [Q&A] [answered]
Hi Thomas,

So, I am back on mfront and CAST3M after working on another part of my PhD. I am trying to implement the EKMC modelled from last time (https://github.com/thelfer/tfel/discussions/462) in CAST3M but I am having troubles with making the computation work:
![image](https://github.com/thelfer/tfel/assets/65529689/4f520ea2-d18f-4b4b-8755-ddb55a2af7b7)

It seems to be an integration issue but I don't get what is happing since the behaviour law is working with an mtest file. Actually, I am not sure about the way I did to implement it in CAST3M. For example, I have an extra internal state variable "ta" but I don't know how to initialize it. I have tried several things with "TAB1 . 'VARIABLES_INTERNES' . 0" in PASAPAS but it seems to not help.

Also, I tried with the same model, to implement a simple Voce law from mfront and it was working so I guess It is coming from the specificities of the law I am trying to implement.

You can find all the files here:
https://github.com/RonanRiverie/Traction/tree/main/EKMC

Thank you very much for your help
Ronan

**Comment by thelfer** [comment]
Hi Ronan,

Please find attached a sligthly modified version. I shifted `ta` to make the initial value egal to 1000 (This shall be done in the Cast3M file, but I forgot how to do it :)).

If I am not mistakend, compared to the MTest file, the imposed strain is much higher in the Cast3M file and the time steps much greater. Also, the material properties are not consistent.

Could make the Cast3M file and the MTest file match ? Beware to change the initial value of `ta` in the MTest file

[ekmc.zip](https://github.com/user-attachments/files/15807709/ekmc.zip)

**Reply by RonanRiverie** [reply]
Actually the problem was well the time step which was too big. I was so focused on the connection between the mfront law and CAST3M because of many errors before, that I forgot to check the time step, thanks.

So, I made both mtest and CAST3M properties be consistent and I indeed found the same results:
![image](https://github.com/thelfer/tfel/assets/65529689/762e903e-5a2b-40ff-af96-ce0d97c9c6b9)

Regarding initialiazing the sate variable "ta", so I just used your code and declare "ta0" as a new material property to be able to declare it directly in CAST3M. 
About that, I saw something in an example from the [CAST3M website](https://www-cast3m.cea.fr/index.php?page=exemples_classes&exemple=gatt_3d):
![image](https://github.com/thelfer/tfel/assets/65529689/7d7fefe3-a70d-4ed8-aa37-53db98dab5fa)
It seems that they use this variable "VAR01" to initialiaze their state variable with the porosity but I tried something similar with the state variable "ta" but I dont' really know what is the exact way of defining it so it was not working:
![image](https://github.com/thelfer/tfel/assets/65529689/b5ab1bb2-89b0-4576-8848-147845a5ab3d)

**Reply by thelfer** [reply]
try this:

```
*****************************************
* INITIALISATION DES VARIABLES INTERNES *
*****************************************

TA_0 = 10000 ;
VAR_CHAMP_0 = ZERO MO_EP 'VARINTER' ;
VAR_1 = 'MANU' 'CHML' MO_EP 'TA' TA_0 
        'TYPE' 'VARIABLES_INTERNES' 'STRESSES';
VAR_0 = VAR_CHAMP_0 '+' VAR_1;
```

**Reply by thelfer** [reply]
try this:

```
*****************************************
* INITIALISATION DES VARIABLES INTERNES *
*****************************************

TA_0 = 10000 ;
VAR_CHAMP_0 = ZERO MO_EP 'VARINTER' ;
VAR_1 = 'MANU' 'CHML' MO_EP 'TA' TA_0 
        'TYPE' 'VARIABLES_INTERNES' 'STRESSES';
VAR_0 = VAR_CHAMP_0 '+' VAR_1;
```

**Reply by thelfer** [reply, marked as answer]
Also, add this line in your MFront file:

```
@ModellingHypotheses {".+"};
```

This will add proper support for plane stress through the `StandardElasticity` brick. Without this, plane stress is supported by the Cast3M interface through some external loop around the behaviour integration, which has a net impact on performances:

- with `@ModellingHypotheses {".+"}` : 41s
- without  `@ModellingHypotheses {".+"}` : 1min29s

Beware that this option changes the order of the state variables:

```
 statev = 'MOTS' 'EEXX' 'EEYY' 'EEZZ' 'EEXY' 'P' 'ETOZ' 'TA';
```

**Reply by thelfer** [reply]
Finally, the jacobian seems quite easy to compute in your case, that would really speed-up your computations.

**Reply by RonanRiverie** [reply]
Yes indeed, the trick was the quotes around the '+' which were missing. So thanks to that I could directly initialize the state variable in CAST3M without creating the material property. 

Ok for the plane stress, so I also tried that and it works fine !

Indeed for the Jacobian matrix this is in my plans.

Thanks a lot Thomas.

**Reply by thelfer** [reply]
> Yes indeed, the trick was the quotes around the '+' which were missing. 

Nope. You simply used `ET`, not `+`

> Indeed for the Jacobian matrix this is in my plans.

Great.
